src/rosinstall_generator/generator.py

```diff
@@ -87,20 +87,27 @@ def _classify_repo_names(distro_name, repo_names):
     return names, unknown_names
 
 
-def _get_packages_for_repos(distro_name, repo_names):
+def _get_packages_for_repos(distro_name, repo_names, source=False):
     package_names = set([])
     unreleased_repo_names = set([])
     wet_distro = get_wet_distro(distro_name)
     for repo_name in repo_names:
-        release_repo = wet_distro.repositories[repo_name].release_repository
-        if release_repo:
-            package_names.update(release_repo.package_names)
+        if source:
+            source_package_xmls = wet_distro.get_source_repo_package_xmls(repo_name)
+        if source and source_package_xmls:
+            source_package_names = source_package_xmls.keys()
+            source_package_names.remove("_ref")
```

Comment:
Can you please add a comment above this line why it is necessary. It took quite a bit to figure it out when reading the patch.

Reply:
Added, also filed some related tickets against `rosdistro` which would help this in the future:

https://github.com/ros-infrastructure/rosdistro/issues/115

https://github.com/ros-infrastructure/rosdistro/issues/116